paimon-format/src/main/java/org/apache/paimon/format/avro/FieldReaderFactory.java

```diff
@@ -483,24 +526,27 @@ public RowReader(Schema schema, List<DataType> fieldTypes, int[] projection) {
         public InternalRow read(Decoder decoder, Object reuse) throws IOException {
             GenericRow row;
             if (reuse instanceof GenericRow
-                    && ((GenericRow) reuse).getFieldCount() == projection.length) {
+                    && ((GenericRow) reuse).getFieldCount() == mapping.length) {
                 row = (GenericRow) reuse;
             } else {
-                row = new GenericRow(projection.length);
+                row = new GenericRow(mapping.length);
             }
 
+            Object[] values = new Object[fieldReaders.length];
             for (int i = 0; i < fieldReaders.length; i += 1) {
-                int[] columns = mapping[i];
-                FieldReader reader = fieldReaders[i];
-                if (columns == null) {
-                    reader.skip(decoder);
+                if (mappingBack[i] >= 0) {
+                    values[i] = fieldReaders[i].read(decoder, row.getField(mappingBack[i]));
                 } else {
-                    Object value = reader.read(decoder, row.getField(columns[0]));
-                    for (int column : columns) {
-                        row.setField(column, value);
-                    }
+                    fieldReaders[i].skip(decoder);
                 }
             }
+
+            values = VectorMappingUtils.mappingObjects(values, mapping);
```

Comment:
It is weird, there is a for loop to values. We don't need this `mappingObjects`.

Reply:
Fixed this.